spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,126 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.actions;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.datasketches.theta.Sketch;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.puffin.Blob;
+import org.apache.iceberg.puffin.StandardBlobTypes;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.spark.SparkReadOptions;
+import org.apache.iceberg.spark.SparkValueConverter;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import scala.Tuple2;
+
+public class NDVSketchGenerator {
+
+  private NDVSketchGenerator() {}
+
+  public static final String APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY = "ndv";
+
+  static List<Blob> generateNDVSketchesAndBlobs(
+      SparkSession spark, Table table, Snapshot snapshot, Set<String> columnsToBeAnalyzed) {
+    Map<Integer, ThetaSketchJavaSerializable> columnToSketchMap =
+        computeNDVSketches(spark, table, snapshot.snapshotId(), columnsToBeAnalyzed);
+    return generateBlobs(table, columnsToBeAnalyzed, columnToSketchMap, snapshot);
+  }
+
+  private static List<Blob> generateBlobs(
+      Table table,
+      Set<String> columns,
+      Map<Integer, ThetaSketchJavaSerializable> sketchMap,
+      Snapshot snapshot) {
+    return columns.stream()
+        .map(
+            columnName -> {
+              Types.NestedField field = table.schema().findField(columnName);
+              Sketch sketch = sketchMap.get(field.fieldId()).getSketch();
+              long ndv = (long) sketch.getEstimate();
+              return new Blob(
+                  StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                  ImmutableList.of(field.fieldId()),
+                  snapshot.snapshotId(),
+                  snapshot.sequenceNumber(),
+                  ByteBuffer.wrap(sketch.toByteArray()),
+                  null,
+                  ImmutableMap.of(APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY, String.valueOf(ndv)));
+            })
+        .collect(Collectors.toList());
+  }
+
+  private static Map<Integer, ThetaSketchJavaSerializable> computeNDVSketches(
```

Comment:
Have we considered using `TypedImperativeAggregate` or any other API that would work with `InternalRow`? I am concerned about the serialization cost from `InternalRow` to `Row`.

I'll have to look into it. @karuppayya, what options did we explore for computing this?

Reply:
I derived the logic from [Datasketches example](https://datasketches.apache.org/docs/Theta/ThetaSparkExample.html). To use the TypedImperativeAggregate shouldn't be a expression natively provided by Spark?
Also can this be taken as improvement in a subsequent PR?

Reply:
> To use the TypedImperativeAggregate shouldn't be a expression natively provided by Spark?

Not necessarily, it is fairly easy to convert `TypedImperativeAggregate` to `Column` and use it. 

```
val func = new ThethaSketchAggregator(...)
new Column(func.toAggregateExpression(isDistinct = false))
```

> Also can this be taken as improvement in a subsequent PR?

I think the performance difference will be significant and I doubt the existing approach will work if Kryo is enabled. The current solution relies on writeObject/readObject methods that will be simply ignored by Kryo.

It will be fairly easy to implement `TypedImperativeAggregate` in our case. We will have to return a row with NDV and bytes for the blob from the `eval` method. We will create a separate aggregate for each column that is being analyzed. Happy to talk more about this offline, if needed.